plugin/evm/epochs_manager_test.go

```diff
@@ -19,57 +19,114 @@ func (t *testEpochs) Details(epoch uint64) (EpochDetails, error) {
 }
 
 func TestByTimestamp(t *testing.T) {
-	callsCount := 0
-	testEpochsMock := &testEpochs{
-		DetailsFunc: func(epoch uint64) (EpochDetails, error) {
-			callsCount++
-			epochDetails1 := EpochDetails{
-				StartTime: big.NewInt(time.Date(2022, time.January, 1, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
-				EndTime:   big.NewInt(time.Date(2022, time.January, 2, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
-			}
-			epochDetails2 := EpochDetails{
-				StartTime: big.NewInt(time.Date(2022, time.January, 2, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
-				EndTime:   big.NewInt(time.Date(2022, time.January, 3, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
-			}
-			epochDetails3 := EpochDetails{
-				StartTime: big.NewInt(time.Date(2022, time.January, 3, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
-				EndTime:   big.NewInt(time.Date(2022, time.January, 4, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
-			}
-			epochDetails4 := EpochDetails{
-				StartTime: big.NewInt(time.Date(2022, time.January, 4, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
-				EndTime:   big.NewInt(time.Date(2022, time.December, 4, 0, 0, 0, 0, time.UTC).Unix()).Uint64(),
-			}
-			if epoch < 0 {
-				return EpochDetails{}, fmt.Errorf("Invalid epoch")
-			}
-			switch epoch {
-			case 0:
-				return epochDetails1, nil
-			case 1:
-				return epochDetails2, nil
-			case 2:
-				return epochDetails3, nil
-			default:
-				return epochDetails4, nil
-			}
-		},
-	}
+	epochs := fakeEpochs(5)
+
+	t.Run("nominal case", func(t *testing.T) {
+		callsCount := 0
+		mock := &testEpochs{
+			DetailsFunc: func(epoch uint64) (EpochDetails, error) {
+				callsCount++
+				return epochs[int(epoch)], nil
+			},
+		}
+
+		epochManager := NewEpochsManager(mock)
+
+		// Request last known epoch. Expect 4 iterations forwards before we reach it, and therefore 4 calls to Details.
+		epoch, err := epochManager.ByTimestamp(epochs[4].StartTime)
+		require.NoError(t, err)
+		assert.Equal(t, uint64(3), epoch)
+		assert.Equal(t, 4, callsCount)
+
+		// Request first epoch. Expect 4 iteration backwards this time, so a total of 8 calls now.
+		epoch, err = epochManager.ByTimestamp(epochs[0].StartTime)
+		require.NoError(t, err)
+		assert.Equal(t, uint64(0), epoch)
+		assert.Equal(t, 8, callsCount)
+
+		// Request central epoch. Expect 2 more iterations forwards.
+		epoch, err = epochManager.ByTimestamp(epochs[2].StartTime)
+		require.NoError(t, err)
+		assert.Equal(t, uint64(1), epoch)
+		assert.Equal(t, 10, callsCount)
+	})
+
+	t.Run("request first epoch", func(t *testing.T) {
```

Comment:
I am wondering in which kind of scenario, the "nominal case" will work but this test will fail.

Reply:
There is indeed some overlap between the tests, but it doesn't matter much IMO. We want to make sure that requesting each of the boundaries works (first and last epoch) as well as intermediate values and going back and forth (nominal case).